Neuzz integration

ImanHosseini opened the pull request:
Build succeeds for all benchmarks except for arrow.

ImanHosseini commented:
@googlebot I signed it!

inferno-chromium commented on fuzzers/neuzz/fuzzer.py:
```diff
@@ -0,0 +1,146 @@
+    """Set environment variables used to build targets for AFL-based
+    fuzzers."""
+    cflags = ['-O2', '-fno-omit-frame-pointer']
+    # ASAN?
```
This is problematic, can we not use existing flags. Setting your custom flags would mean it wont work with bug based benchmarks (they use ASan). It is better to remove/change part of flags that dont work (e.g. optimization flag).

ImanHosseini replied:
done,

inferno-chromium commented:
Thank you so much for bringing in Neuzz, we are very excited!

lszekeres commented:
Thanks so much, Iman! 

Are there special requirements to run this? Do we need GPUs?

inferno-chromium commented:
Need to run `make format` to clean up lint issues. Can verify everything is fine with `make presubmit`

ImanHosseini commented on fuzzers/neuzz/fuzzer.py:
```diff
@@ -0,0 +1,146 @@
+                            stdout=subprocess.DEVNULL,
+                            stderr=subprocess.DEVNULL)
+
+    # subprocess.check_call(command, stdout=output_stream, stderr=output_stream)
```
done. (except the line for custom flags, can we keep that? It's useful to comment/uncomment that to test)

inferno-chromium replied:
yes can keep comments.

ImanHosseini commented on fuzzers/neuzz/fuzzer.py:
```diff
@@ -0,0 +1,146 @@
+              hide_output=False):
+    """Run afl-fuzz."""
+    # Spawn the afl fuzzing process.
+    print('[run_neuzz] Running target with afl-fuzz')
```
done.

ImanHosseini commented on fuzzers/neuzz/fuzzer.py:
```diff
@@ -0,0 +1,146 @@
+    shutil.copy('/neuzz/nn.py', output_directory)
+
+
+def prepare_fuzz_environment(input_corpus):
```
done.

ImanHosseini commented on fuzzers/neuzz/runner.Dockerfile:
```diff
@@ -0,0 +1,24 @@
+    python --version && \
+    python -m pip install --upgrade pip==20.3 && \
+    python -m pip install tensorflow==1.8.0 && \
+    python -m pip install keras==2.2.3
```
done.

ImanHosseini commented:
@inferno-chromium Sorry for the lint issues, I ran make format and it printed 'Success' but it wasn't fixing anything, I fixed the 'presubmit' lint errors manually in the end. Now presubmit succeeds (at least that's what I see) but spews 6 warnings which are regarding other code not related to neuzz.

@lszekeres as far as I know, since tensorflow 1.15 tensorflow-gpu is not separate, so if this would run with a gpu it would use it, otherwise fallback to run on cpu. Whether how that would impact the performance, I'm not really sure. Basically I took the old neuzz integration and updated it to work. I also haven't ran a day-long experiment for example, to see whether it crashes or something, but based on the original descriptions in the neuzz paper and repo, this should be the way to go.

ImanHosseini commented on fuzzers/neuzz/fuzzer.py:
```diff
@@ -0,0 +1,146 @@
+    command = [
+        "cp", "-RT", f"{output_corpus}/queue/", f"{input_corpus}_neuzzin/"
+    ]
+    print('[run_neuzz] Running command: ' + ' '.join(command))
```
done.

ImanHosseini reviewed (commented):
Applied the changes.

inferno-chromium reviewed (approved):
AMAZING, looks like CI worked.
